### ISSUE TITLE
Merge lora permanently?

### ISSUE THREAD

**Issue by Ph0rk0z** [johnsmith0031/alpaca_lora_4bit#12]
Is it possible to save the merged PT file like it is for stable diffusion? Would it still need PEFT and all this other stuff?

**Comment by johnsmith0031** [comment]
I think it is hard to do because the 4-bit model is already quantized once and if you reconstruct it and add lora-weight to the weight and quantize it again the loss may be out of control. But technically feasible, if ignore the quality of the merged model.

**Comment by johnsmith0031** [comment]
Yes then I think it's feasible.

**Comment by ehartford** [comment]
I wrote a script for that, but the resulting model won't fit in VRAM.  (30b anyway)

```
import sys
sys.path.insert(0, './repository/transformers/src')
sys.path.insert(0, './repository/GPTQ-for-LLaMa')
sys.path.insert(0, './repository/peft/src')

import peft
import peft.tuners.lora
assert peft.tuners.lora.is_gptq_available()

import torch
import transformers
from autograd_4bit import load_llama_model_4bit_low_ram
from peft import LoraConfig, get_peft_model, get_peft_model_state_dict, PeftModel

print('PyTorch version:', torch.__version__)
print('Transformers version:', transformers.__version__)
print('PEFT version:', peft.__version__)
print('cuda:', torch.cuda.is_available())

# ! Config
from arg_parser import get_config
import train_data

ft_config = get_config()

# * Show loaded parameters
if ft_config.local_rank == 0:
    print(f"{ft_config}\n")

if ft_config.gradient_checkpointing:
    print('Disable Dropout.')

# Load Basic Model
model, tokenizer = load_llama_model_4bit_low_ram(ft_config.llama_q4_config_dir, ft_config.llama_q4_model, device_map=ft_config.device_map)

# Load base model and adapter
adapter_path = './alpaca_lora/adapter_model.bin'
adapter_weights = torch.load(adapter_path)

# Merge the adapter weights into the base model
for name, param in model.named_parameters():
    if name in adapter_weights:
        param.data = adapter_weights[name]

# Save the merged model
output_path = './merged_model'
model.save_pretrained(output_path)
```

**Comment by LoopControl** [comment]
The above script doesn't seem to work after the v2 GPTQ changes -- it produces a model file that is exactly the same as the original.

Anyone know how to fix (it looks like the layer names don't match for that `for name,param` loop)?
